### PR TITLE
Add vm.buffer.hash and util.buffer.hash ops

### PR DESCRIPTION
**Pull request by qedawkins**
This adds support for computing the SipHash24 of a byte range of a buffer, with the goal of using this as a primitive for building a hash table out of VM IR.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Util/IR/test/buffer_ops.mlir [hunk 1]
```diff
@@ -126,3 +126,14 @@ func.func @buffer_store(%arg0: !util.buffer, %arg1: index, %arg2: i32) {
...
+// -----
+
+// CHECK-LABEL: @buffer_hash
+func.func @buffer_hash(%arg0: !util.buffer, %arg1: index) -> i64 {
+  %c17 = arith.constant 17 : index
+  %c100 = arith.constant 100 : index
+  // CHECK: = util.buffer.hash %arg0[%c100 for %c17] : !util.buffer{%arg1} -> i64
+  %0 = util.buffer.hash %arg0[%c100 for %c17] : !util.buffer{%arg1} -> i64
+  return %0 : i64
+}
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/IR/test/buffer_ops.mlir]: missing newline at end of file in this and other .mlir files (bad editor setting?)

File: runtime/src/iree/vm/BUILD.bazel [hunk 2]
```diff
@@ -78,6 +78,7 @@ iree_runtime_cc_library(
     ],
     deps = [
         "//runtime/src/iree/base",
+        "//runtime/src/iree/base:core_headers",
```
Comment (benvanik) [inline runtime/src/iree/vm/BUILD.bazel]: base pulls in what's needed - you can remove the #include of alignment.h and this dep

File: runtime/src/iree/vm/buffer.c [hunk 3]
```diff
@@ -314,6 +315,90 @@ IREE_API_EXPORT iree_status_t iree_vm_buffer_write_elements(
...
+
+  v0 ^= hash;
+  v2 ^= 0xff;
+
+  for (int i = 0; i < dROUNDS; ++i) {
+    SIPROUND(v0, v1, v2, v3);
+  }
+
+  hash = v0 ^ v1 ^ v2 ^ v3;
+  iree_unaligned_store_le_u64(result, hash);
```
Comment (benvanik) [inline runtime/src/iree/vm/buffer.c]: no need for the unaligned store as the result pointer is always aligned

File: runtime/src/iree/vm/buffer.c [hunk 4]
```diff
@@ -314,6 +315,90 @@ IREE_API_EXPORT iree_status_t iree_vm_buffer_write_elements(
...
+#ifndef dROUNDS
+#define dROUNDS 4
+#endif
+
+IREE_API_EXPORT iree_status_t iree_vm_buffer_hash(
+    const iree_vm_buffer_t* source_buffer, iree_host_size_t source_offset,
+    iree_host_size_t length, int64_t* result) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_ASSERT_ARGUMENT(source_buffer);
+  // Get the byte span for the source data.
```
Comment (benvanik) [inline runtime/src/iree/vm/buffer.c:350]:
```suggestion

  // Get the byte span for the source data.
```

File: runtime/src/iree/vm/buffer.h [hunk 5]
```diff
@@ -199,6 +199,12 @@ IREE_API_EXPORT iree_status_t iree_vm_buffer_write_elements(
     iree_host_size_t target_offset, iree_host_size_t element_count,
     iree_host_size_t element_length);
 
+// Computes the SipHash-2-4 of the source buffer along the byte range starting
+// as |source_offset| of |length| bytes. Uses 128-bit key `0x00010203...0e0f`.
+IREE_API_EXPORT iree_status_t iree_vm_buffer_hash(
+    const iree_vm_buffer_t* source_buffer, iree_host_size_t source_offset,
+    iree_host_size_t length, int64_t* result);
```
Comment (benvanik) [inline runtime/src/iree/vm/buffer.h:206]: if this is output only then prefix with `out_` - if it's I/O then just `result` is fine but we'd want to document the expectations (see comment about incremental hashing).

```suggestion
    iree_host_size_t length, int64_t* out_result);
```

File: runtime/src/iree/vm/buffer.c [hunk 6]
```diff
@@ -314,6 +315,90 @@ IREE_API_EXPORT iree_status_t iree_vm_buffer_write_elements(
   return iree_ok_status();
 }
 
+// Based on reference implementation from https://github.com/veorq/SipHash
+// By Jean-Philippe Aumasson and Daniel J. Bernstein.
```
Comment (benvanik) [inline runtime/src/iree/vm/buffer.c]: worth noting here that it's CC0-licensed (effectively public domain)

**Review by benvanik — CHANGES_REQUESTED** [review]
Minor style nits and such,

It'd be nice to know how we could make this incremental (an optional state buffer, etc) but that can come with some larger crypto module that does more stuff than what this is intended for (basic hashing of simple bytes spans). Keeping it simple here means we can optimize it pretty tightly/port it pretty easily!

**Comment by qedawkins** [comment]
> Minor style nits and such,
> 
> It'd be nice to know how we could make this incremental (an optional state buffer, etc) but that can come with some larger crypto module that does more stuff than what this is intended for (basic hashing of simple bytes spans). Keeping it simple here means we can optimize it pretty tightly/port it pretty easily!

I found this investigation of a few different choices for incremental hash functions so maybe we can import learnings from there: https://github.com/rust-lang/rust/issues/41215

**Review by benvanik — APPROVED** [review]

**Pull request merged**